spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java

```diff
@@ -124,16 +125,26 @@ public void setJobPollInterval(Duration jobPollInterval) {
 	@Override
 	public ListenableFuture<Job> writeDataToTable(
 			String tableName, InputStream inputStream, FormatOptions dataFormatOptions) {
+		return this.writeDataToTable(tableName, inputStream, dataFormatOptions, null);
+	}
+
+	@Override
+	public ListenableFuture<Job> writeDataToTable(
+			String tableName, InputStream inputStream, FormatOptions dataFormatOptions, Schema schema) {
+
 		TableId tableId = TableId.of(datasetName, tableName);
 
-		WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
+		WriteChannelConfiguration.Builder writeChannelConfiguration = WriteChannelConfiguration
 				.newBuilder(tableId)
 				.setFormatOptions(dataFormatOptions)
-				.setAutodetect(this.autoDetectSchema)
 				.setWriteDisposition(this.writeDisposition)
-				.build();
+				.setAutodetect(this.autoDetectSchema);
 
-		TableDataWriteChannel writer = bigQuery.writer(writeChannelConfiguration);
+		if (schema != null) {
+			writeChannelConfiguration.setSchema(schema);
```

Comment:
Is this compatible with auto-detecting schema?
What happens if schema is not provided, and auto-detection is off?

Reply:
Yes good question - the behavior is like this:

With auto-detect schema on:
* If schema is provided - it does its best to "fit" the data to the provided schema. For example, if you provided 3 columns in the schema, and the dataset contains extra (more than 3) columns, it will only extract the 3 columns you specified from the dataset.
* If schema is not provided - It will simply infer schema from the dataset

With auto-detect schema off:
* If schema is provided - Use the explicitly provided schema to create table for dataset. This is _strict_, so if the schema specifies a different number of columns than the dataset it will fail. It will not make a best effort attempt to fit the data.
* If schema is not provided - This will simply fail.